executor/pipelined_window.go

```diff
@@ -0,0 +1,479 @@
+	p.rows = append(p.rows, rows...)
+	rows = rows[:0]
+	if p.end.UnBounded {
+		if !p.whole {
+			// we can't proceed until the whole partition is consumed
+			return true, nil
+		}
+	}
+	// let's just try produce
+	return false, nil
+}
+
+// finish is called upon a whole partition is consumed
+func (p *processor) finish() {
+	p.whole = true
+}
+
+func (p *processor) getStart(ctx sessionctx.Context) (uint64, error) {
+	if p.start.UnBounded {
+		return 0, nil
+	}
+	if p.isRangeFrame {
+		var start uint64
+		for start = p.lastStartRow; start < p.rowCnt; start++ {
+			var res int64
+			var err error
+			for i := range p.orderByCols {
+				res, _, err = p.start.CmpFuncs[i](ctx, p.orderByCols[i], p.start.CalcFuncs[i], p.getRow(start), p.getRow(p.curRowIdx))
+				if err != nil {
+					return 0, err
+				}
+				if res != 0 {
+					break
+				}
+			}
+			// For asc, break when the calculated result is greater than the current value.
+			// For desc, break when the calculated result is less than the current value.
+			if res != p.expectedCmpResult {
+				break
+			}
+		}
+		return start, nil
+	}
+	switch p.start.Type {
+	case ast.Preceding:
+		if p.curRowIdx > p.start.Num {
+			return p.curRowIdx - p.start.Num, nil
+		}
+		return 0, nil
+	case ast.Following:
+		return p.curRowIdx + p.start.Num, nil
+	default: // ast.CurrentRow
+		return p.curRowIdx, nil
+	}
+}
+
+func (p *processor) getEnd(ctx sessionctx.Context) (uint64, error) {
+	if p.end.UnBounded {
+		return p.rowCnt, nil
+	}
+	if p.isRangeFrame {
+		var end uint64
+		for end = p.lastEndRow; end < p.rowCnt; end++ {
+			var res int64
+			var err error
+			for i := range p.orderByCols {
+				res, _, err = p.end.CmpFuncs[i](ctx, p.end.CalcFuncs[i], p.orderByCols[i], p.getRow(p.curRowIdx), p.getRow(end))
+				if err != nil {
+					return 0, err
+				}
+				if res != 0 {
+					break
+				}
+			}
+			// For asc, break when the calculated result is greater than the current value.
+			// For desc, break when the calculated result is less than the current value.
+			if res == p.expectedCmpResult {
+				break
+			}
+		}
+		return end, nil
+	}
+	switch p.end.Type {
+	case ast.Preceding:
+		if p.curRowIdx >= p.end.Num {
+			return p.curRowIdx - p.end.Num + 1, nil
+		}
+		return 0, nil
+	case ast.Following:
+		return p.curRowIdx + p.end.Num + 1, nil
+	default: // ast.CurrentRow:
+		return p.curRowIdx + 1, nil
+	}
+}
+
+// produce produces rows and append it to chk, return produced means number of rows appended into chunk, available means
+// number of rows processed but not fetched
+func (p *processor) produce(ctx sessionctx.Context, chk *chunk.Chunk, remained int) (produced int64, err error) {
+	var (
+		start uint64
+		end   uint64
+		more  bool
+	)
+	for remained > 0 {
+		more, err = p.moreToProduce(ctx)
+		if err != nil {
+			return
+		}
+		if !more {
+			break
+		}
+		start, err = p.getStart(ctx)
+		if err != nil {
+			return
+		}
+		end, err = p.getEnd(ctx)
+		if err != nil {
+			return
+		}
```

Comment:
Can we use moreToProduce to return start/end? It can avoid duplicated calculation...

Reply:
moreToProduce won't caculate start/end all the time, and it will make the code dirty